File: neetcode150/arrays_and_hashing/leetcode271.py
from typing import List


class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string.
        """
        encoded = []

        for s in strs:
            encoded.append(f"{len(s)}/{s}")

        return "".join(encoded)

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings.
        """
        decoded = []
        start = 0

        while start < len(s):
            slash = s.find("/", start)
            # we can assume that this will always exist
            s_len = int(s[start: slash])

            decoded.append(s[slash + 1: slash + 1 + s_len])

            start = slash + 1 + s_len

        return decoded

File: js_algorithms/neetcode150/array_and_hashing/leetcode271.js
/**
 * Encodes a list of strings to a single string.
 *
 * @param {string[]} strs
 * @return {string}
 */
const encode = function (strs) {
  return strs.map((str) => `${str.length}/${str}`).join('');
};

/**
 * Decodes a single string to a list of strings.
 *
 * @param {string} s
 * @return {string[]}
 */
const decode = function (s) {
  const result = [];
  let start = 0;

  while (start < s.length) {
    const slash = s.indexOf('/', start);
    const length = parseInt(s.substring(start, slash));
    const subResult = s.substring(slash + 1, slash + length + 1);

    result.push(subResult);
    start = slash + length + 1;
  }

  return result;
};

module.exports = { encode, decode };
